python/ray/data/_internal/arrow_serialization.py

```diff
@@ -63,3 +101,199 @@ def _register_arrow_json_parseoptions_serializer():
+        return []
+
+    from ray.data.extensions import ArrowTensorArray, ArrowVariableShapedTensorArray
+
+    array_types = [
+        pa.lib.NullArray,
+        pa.lib.BooleanArray,
+        pa.lib.UInt8Array,
+        pa.lib.UInt16Array,
+        pa.lib.UInt32Array,
+        pa.lib.UInt64Array,
+        pa.lib.Int8Array,
+        pa.lib.Int16Array,
+        pa.lib.Int32Array,
+        pa.lib.Int64Array,
+        pa.lib.Date32Array,
+        pa.lib.Date64Array,
+        pa.lib.TimestampArray,
+        pa.lib.Time32Array,
+        pa.lib.Time64Array,
+        pa.lib.DurationArray,
+        pa.lib.HalfFloatArray,
+        pa.lib.FloatArray,
+        pa.lib.DoubleArray,
+        pa.lib.ListArray,
+        pa.lib.LargeListArray,
+        pa.lib.MapArray,
+        pa.lib.FixedSizeListArray,
+        pa.lib.UnionArray,
+        pa.lib.BinaryArray,
+        pa.lib.StringArray,
+        pa.lib.LargeBinaryArray,
+        pa.lib.LargeStringArray,
+        pa.lib.DictionaryArray,
+        pa.lib.FixedSizeBinaryArray,
+        pa.lib.Decimal128Array,
+        pa.lib.Decimal256Array,
+        pa.lib.StructArray,
+        pa.lib.ExtensionArray,
+        ArrowTensorArray,
+        ArrowVariableShapedTensorArray,
+    ]
+    try:
+        array_types.append(pa.lib.MonthDayNanoIntervalArray)
+    except AttributeError:
+        # MonthDayNanoIntervalArray doesn't exist on older pyarrow versions.
+        pass
+    return array_types
+
+
+def _arrow_array_reduce(a: "pyarrow.Array"):
+    """Custom reducer for Arrow arrays that works around a zero-copy slicing pickling
+    bug by using the Arrow IPC format for the underlying serialization.
+    """
+    from pyarrow.ipc import RecordBatchStreamWriter
+    from pyarrow.lib import RecordBatch, BufferOutputStream
+
+    batch = RecordBatch.from_arrays([a], [""])
+    output_stream = BufferOutputStream()
+    with RecordBatchStreamWriter(output_stream, schema=batch.schema) as wr:
+        wr.write_batch(batch)
+    return _restore_array, (output_stream.getvalue(),)
+
+
+def _restore_array(buf: bytes) -> "pyarrow.Array":
+    from pyarrow.ipc import RecordBatchStreamReader
+
+    with RecordBatchStreamReader(buf) as reader:
+        return reader.read_next_batch().column(0)
+
+
+def _arrow_chunkedarray_reduce(a: "pyarrow.ChunkedArray"):
+    """Custom reducer for Arrow ChunkedArrays that works around a zero-copy slicing
+    pickling bug by using the Arrow IPC format for the underlying serialization.
+    """
+    import pyarrow as pa
+    from pyarrow.ipc import RecordBatchStreamWriter
+    from pyarrow.lib import RecordBatch, BufferOutputStream
+
+    # Convert chunked array to contiguous array.
+    if a.num_chunks == 0:
+        a = pa.array([], type=a.type)
+    elif isinstance(a.type, pa.ExtensionType):
+        chunk = a.chunk(0)
+        a = type(chunk).from_storage(
+            chunk.type, pa.concat_arrays([c.storage for c in a.chunks])
+        )
+    else:
+        a = a.combine_chunks()
```

Comment:
This seems to need memory allocation and no longer zero-copy. Can it serialize each chunk to a output without moving chunks into contiguous region?

Reply:
Yep I had a TODO for finding a way around this, turns out the `RecordBatchStreamWriter.write_table()` API will write each underlying horizontal set of column chunks as a record batch, so it will be zero-copy. I just ported the `ChunkedArray` serialization path to that API!